examples/keras_recipes/float8_training_and_inference_with_transformer.py

```diff
@@ -0,0 +1,267 @@
+Typically, E4M3 is best used during the forward pass because activations and
+weights require more precision. In the backward pass, however, E5M2 is utilized
+because gradients are less susceptible to the loss of precision but require
+higher dynamic range.
+
+It is worth noting that FP8 inference deployment is greatly simplified, as
+inference and training use the same datatype. This is in contrast to INT8
+inference with networks trained in 32- or 16-bit floating point, which require
+post-training quantization (PTQ) calibration and even quantization-aware
+training (QAT) in order to maintain model accuracy.
+
+In this example, we will build a simple Transformer model and train it with
+both FP16 and FP8 precision. You will observe that the accuracy doesn't decrease
+with lower precision.
+
+Note: You will need a decent GPU with FP8 Tensor Cores support for the expected
+performance improvement.
+"""
+
+"""
+## Setup
+
+We will use KerasNLP library to simplify the model implementation. Additionally,
+use mixed precision training to reduce the training time.
+"""
+
+"""shell
+pip install -q --upgrade keras-nlp
+pip install -q --upgrade keras  # Upgrade to Keras 3.
+"""
+
+import os
+
+os.environ["KERAS_BACKEND"] = "jax"
+
+import re
+
+import keras
+import keras_nlp
+
+keras.config.set_dtype_policy("mixed_bfloat16")
+
+"""
+## Dataset
+
+Let's load the IMDB dataset from `keras.datasets`. This dataset consists of
+25,000 movie reviews labeled by sentiment (positive/negative). To simplify the
+task, we only consider the top 20k words and the first 200 words of each review.
+"""
+
+vocabulary_size = 20000
+max_sequence_length = 200
+index_from = 3
+(x_train, y_train), (x_val, y_val) = keras.datasets.imdb.load_data(
+    num_words=vocabulary_size, start_char=1, oov_char=2, index_from=3
+)
+print(f"Number of training sequences: {len(x_train)}")
+print(f"Number of validation sequences: {len(x_val)}")
+x_train = keras.utils.pad_sequences(x_train, maxlen=max_sequence_length)
+x_val = keras.utils.pad_sequences(x_val, maxlen=max_sequence_length)
+
+"""
+To better understand the dataset, we can print one encoded data, its decoded
+sequence, and the label in the training data.
+"""
+
+word_index = keras.datasets.imdb.get_word_index()
+inverted_word_index = dict((i + index_from, word) for (word, i) in word_index.items())
+inverted_word_index[1] = "[START]"
+inverted_word_index[2] = "[OOV]"
+
+decoded_sequence = " ".join(inverted_word_index[i] for i in x_train[0])
+print(f"Encoded training data:\n{x_train[0]}")
+print(f"Decoded training data:\n{decoded_sequence}")
+print(f"Labeld sentiment (0: negative, 1: positive):\n{y_train[0]}")
+
+"""
+## Model
+
+Let's build a simple Transformer model. `TransformerBlock` outputs one vector
+for each time step of our input sequence. Here, we take the mean across all time
+steps and use a feedforward network on top of it to classify text.
+"""
+
+
+class TransformerBlock(keras.layers.Layer):
```

Comment:
You can just grab this class from KerasNLP -- let's keep the code as short as possible, and focused on the float8 training part.

Reply:
Actually, there is an issue in KerasNLP:
https://github.com/keras-team/keras-nlp/pull/1623

TLDR: We need to explicitly expose `training=None` in the signature of `call` in subclasses. This will also fail the float8 training.

Reply:
I've reimplemented the class like this:
```python
class TransformerDecoder(keras_nlp.layers.TransformerDecoder):
    def call(
        self,
        decoder_sequence,
        encoder_sequence=None,
        decoder_padding_mask=None,
        decoder_attention_mask=None,
        encoder_padding_mask=None,
        encoder_attention_mask=None,
        self_attention_cache=None,
        self_attention_cache_update_index=None,
        cross_attention_cache=None,
        cross_attention_cache_update_index=None,
        use_causal_mask=True,
        training=None,
    ):
        return super().call(
            decoder_sequence,
            encoder_sequence,
            decoder_padding_mask,
            decoder_attention_mask,
            encoder_padding_mask,
            encoder_attention_mask,
            self_attention_cache,
            self_attention_cache_update_index,
            cross_attention_cache,
            cross_attention_cache_update_index,
            use_causal_mask,
        )
```

Reply:
I would recommend fixing the issue in KerasNLP first, then relying on plain KerasNLP classes in the example.

Reply:
It should be fixed now. I have updated the example.